stumpy/core.py

```diff
@@ -2885,3 +2885,78 @@ def max_distance(D):
         candidate_idx = np.argmin(D)
 
     return np.array(matches, dtype=object)
+
+
+@njit(
+    # "(i8, i8, i8, i8[:], i8)"
+    fastmath=True,
+    parallel=True,
+)
+def _get_tiles(m, n_A, n_B, diags, tile_size):
```

Comment:
I don't think getting a single tile like that is appropriate, because we're distributing those tiles into different threads, just like how `stump` distributes diagonals into different threads. We need `ndist` counts for the tiles in order to determine which thread gets which tile.

I also feel like `tile_i` and `tile_j` are more consistent with the `i, j` notation we use in `stump`. `tile_row` and `tile_col` make it sound like we're talking about the row and columns of the tiles themselves, which is not the case. Rather, we are referring to the `i` and `j` values of the distance matrix that marks the top left corner of the tile. What do you think?

Reply:
> I don't think getting a single tile like that is appropriate, because we're distributing those tiles into different threads, just like how stump distributes diagonals into different threads. We need ndist counts for the tiles in order to determine which thread gets which tile.

But couldn't we assign the tile indices to each thread and then have each thread figure out its own tiles?

> I also feel like tile_i and tile_j are more consistent with the i, j notation we use in stump. tile_row and tile_col make it sound like we're talking about the row and columns of the tiles themselves, which is not the case. Rather, we are referring to the i and j values of the distance matrix that marks the top left corner of the tile. What do you think?

That's fair. I'm somewhat indifferent. If anything we typically `i` for `T_A` and `j` for `T_B` exclusively so let's make sure that is also the case here. I know we aren't indexing into `T_A` and `T_B` but the distance matrix certainly has a length/width that corresponds to `T_A` and `T_B`

Reply:
> But couldn't we assign a list of tile indices to each thread and then have each thread figure out its own tile info? The benefit is that we don't store all of this somewhat redundant tile info and, instead, we compute it on-the-fly. Right now, for large time series, quite a bit of memory is being consumed and the code is a bit more complex

Ahh I see what you mean. So the only thing we would need before we begin the threads is the `ndist` values for each tile, because `ndist` is used for dividing the tiles into the threads, but you're proposing that the rest of the tile info is computed on-the-fly within each tile? I would almost be okay with that, except that when we compute `ndist`, we need the `tile_start_diag`, `tile_stop_diag`, `tile_width` and `tile_height` values, which makes me ask "if we already have these values, why not just keep them and use later"?

Reply:
> That's fair. I'm somewhat indifferent. If anything we typically i for T_A and j for T_B exclusively so let's make sure that is also the case here. I know we aren't indexing into T_A and T_B but the distance matrix certainly has a length/width that corresponds to T_A and T_B

Makes sense, I will review the logic to make sure we are always consistent with this.

Reply:
> "if we already have these values, why not just keep them and use later"

It depends. In general, I agree with you. However, again, for large time series, I think it would cost a lot less to re-compute things vs storing all of the tile info in memory. It's a trade off. These tiles are basically `6 x n` in shape, which is big when `n` is large. I don't know the right answer but I do see that the code needs further simplification (aside from refactoring with `stump`).

Reply:
@seanlaw good point, not totally sure myself what side of the tradeoff we should be in, I'm okay implementing it either way.

> These tiles are basically `6 x n` in shape, which is big when n is large

It shouldn't be that many because right now each tile holds `6` extra values, and at the worst case scenario, there will be `ceil(n_A / 4096) * ceil(n_B / 4096)` tiles (4096 is the current `STUMPY_MAX_TILE_SIZE` value, although this might change in the next PR). So if `n` is `100,000`, we only have `625` tiles, and if `n` is `1,000,000`, we have `60,000` tiles.